src/main/scala/com/atamaniv/CrimeSupervisor.scala

```diff
@@ -1,15 +1,121 @@
 package com.atamaniv
 
-import akka.actor.{Actor, ActorLogging, Props}
+import akka.actor.{Actor, ActorLogging, PoisonPill, Props}
+import com.atamaniv.ColumnNames.ColumnNames
+import com.atamaniv.Messages._
+import com.atamaniv.model.{Coordinates, Crime}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.{Dataset, Row, SparkSession}
 
-object IotSupervisor {
+import scala.util.Try
+
+/***
+  * CrimeSupervisor class is responsible for lifecycle of FolderReader actor and FileReader actors,
+  * It consumes datasets form fileReaders and prints merged result based of Criteria
+  * */
+
+object CrimeSupervisor {
   def props(): Props = Props(new CrimeSupervisor)
 }
 
+object ColumnNameConversions {
+  implicit def ColumnNamesToString(value: ColumnNames) =
+    value.toString
+}
+
 class CrimeSupervisor extends Actor with ActorLogging {
+  var mainDataSet: Option[Dataset[Row]] = None
+  var expectedFiles = 0
+
   override def preStart(): Unit = log.info("Crime supervisor started")
-  override def postStop(): Unit = log.info("Crime supervisor stopped")
 
-  override def receive = Actor.emptyBehavior
+  override def postStop(): Unit = log.info("**************** Application has been terminated **********************")
+
+  override def receive: Receive = {
+    case StartApplication(path) =>
+      val folderReader = context.actorOf(FolderReader.props(), "ActorOfFolder_" + path.name)
+      folderReader ! GetCsvFiles(path)
+
+    case CsvFiles(files) =>
+      expectedFiles = files.length
+      files.foreach(file => {
+        val fileReader = context.actorOf(FileReader.props(), "ActorOfFile_" + file.name)
+        fileReader ! ReadCsvFile(file.toString())
+      })
+
+    case RawData(dataset) =>
+      expectedFiles -= 1
+      log.info("Merging " + dataset.count() + " rows")
+      mainDataSet = mergeDatasets(dataset)
+      mainDataSet.foreach(ds => log.info(s"Total number of rows: ${ds.count()}"))
+      if (expectedFiles == 0) {
+        self ! SortDataAndPrintTop5
+      }
+
+    case SortDataAndPrintTop5 =>
+      mainDataSet.foreach(crimeList)
+      self ! PoisonPill
+
+    case PrintMessage(message) => println(message)
+  }
+
+  private def mergeDatasets(value: Dataset[Row]): Option[Dataset[Row]] = {
+    mainDataSet match {
+      case Some(dataset) => Some(dataset.union(value))
+      case None => Some(value)
+    }
+  }
+
+  private def crimeList(dataSet: Dataset[Row]): Unit = {
+    val session = getSparkSession()
+    import ColumnNameConversions._
+    import session.implicits._
+    val top5Result = dataSet.map(row => {
+
+      def convertToString(income: Any): Option[String] = {
+        if (income != null) {
+          Some(income.asInstanceOf[String])
+        } else {
+          None
+        }
+      }
+
+      val id = row.getAs(ColumnNames.ID).asInstanceOf[String]
```

Comment:
why not use `val id = row.getAs[String](ColumnNames.ID)`?, instanceOf is usually considered as a code smell)